Raise model index limit

dpsutton opened the pull request:
closes https://github.com/metabase/metabase/issues/40639
closes https://github.com/metabase/metabase/issues/40640

#### Raise the limit of model index values.

We previously only kept 5,000 values from a model. This proved to be too low in practice. On stats we have the following models with the following cardinalities:

| model id | row count |
|  -----------|-----------|
| 10149            | 3983             |
| 15189            | 85               |
| 15130            | 23195            |
| 11654            | 1193             |
| 6736             | 14437            |

This raises the count of rows that we allow to 25,000.

There are a few impacts:
- When diffing the sets, how big of a memory pressure do we have (at most 25,000 strings in a set, and 25,000 strings in another set). This seems fine. Perhaps we should add a substring to the strings. I think this can be follow up work
- When updating the values, make sure our sql statements aren't too large. (uses `(doseq [additions-part (partition-all 10000 additions)] ....)` to ensure we are bounded)
- Search shouldn't bog down. The model index values table has an index `"unique_model_index_value_model_index_id_model_pk" UNIQUE CONSTRAINT, btree (model_index_id, model_pk)` on the model_index_id and primary key but not on the value. This is presumably a table scan and the table has the capability to get much larger. We should watch stats and see how this behaves in practice.

#### Bugs fixed while in there.

This also fixes https://github.com/metabase/metabase/issues/40639. The issue has full repro but the gist is that we used to do this logic:
- get all current values in index
- get all values in the model we want to index
- set difference to find new values, set difference to find old values
- remove any old values (error here)
- add any new values

The error in "remove any old values" was that we used a column on the `model_index` table rather than the column on the `model_index_value` table (`pk_ref` was used instead of `model_pk`). This meant every model index that had values that went away went into an error state. Appropriate tests and a fix are included here.
<img width="1027" alt="image" src="https://github.com/metabase/metabase/assets/6377293/e2ec0772-a087-4b48-b7bf-63419ae38519">

#### After this PR
Every column in the above screenshot should now be "indexed". One is correctly indexed; one is "overflow", meaning it has more than 5,000 values. The others in the "error" state should correctly be indexed, adding new values and removing values no longer present.

dpsutton committed:
Use expression fix on both pk-ref and value-ref

The gist: inside the query expressions are referred to as [:expression
"foo"]. But as a nested query, it's just another field, and the fact
that it was an expression is irrelevant and should not leak from that
stage. So "downstream" should just be [:field "foo" base-type].

Oddly, if you just fix one field or the other, the query would be
"valid", would log an error, but no valid values would be returned.

It seems like the invalid expression reference was just dropped so we
only selected a single column and those were removed (we run `(filter
valid-tuples?)` as we only want valid key/value pairs).

So let's add a `(> (count values) 0)` which was failing previously.

dpsutton committed:
Fix indexing error

When updating the model index, it diffs the existing values vs the new
values and deletes appropriate and adds the new values. But it was using
the wrong column when updating the model index values. It used `pk_ref`
but that table uses `model_pk`. This led to the following errors on
stats:

```sql
select pk_ref, value_ref, state, error from model_index where state = 'error'

pk_ref,value_ref,state,error
"[""field"",563028,null]","[""field"",563018,null]",error,"ERROR: column ""pk_ref"" does not exist Position: 68"
"[""field"",534030,null]","[""field"",534037,null]",error,"ERROR: column ""pk_ref"" does not exist Position: 68"
"[""field"",545945,null]","[""field"",545948,null]",error,"ERROR: column ""pk_ref"" does not exist Position: 68"
```

Now those errors will be fixed and the model indexes will correctly
track the values

dpsutton committed:
Bump indexing threshold to 25,000

Had been limited to 5,000 rows of model indexing. This was a made up
number as we went into this endeavor. But it's proving to be far too
small. Bumping to 25,000

dpsutton committed:
Parititon deletions and inserts

With a higher limit need to ensure that we don't send too large of a
query.

```clojure
model-index=> (t2/insert! ModelIndexValue
                          (map (fn [id]
                                 {:name           (str (gensym "value"))
                                  :model_pk       id
                                  :model_index_id 1})
                               (range 25000)))
Execution error (PSQLException) at org.postgresql.jdbc.PgPreparedStatement/<init> (PgPreparedStatement.java:105).
PreparedStatement can have at most 65,535 parameters. Please consider using arrays, or splitting the query in several ones, or using COPY. Given query has 75,000 parameters
```

johnswanson reviewed (approved):
LGTM, though my knowledge of this area of the code is pretty limited.

dpsutton committed:
move comment and fix test failure explanation

comment should be next to the number? check. and also need to use
`mc/explain` which returns info rather than `mc/validate` which returns
a bool:

```clojure
;; bad
model-index-test=> (let [values [[1 "foo"] [2 "foo"] ["foo" "foo"]]]
                     (-> (mc/validate [:sequential [:tuple number? string?]] values)
                         (me/humanize)))
nil
;; good
model-index-test=> (let [values [[1 "foo"] [2 "foo"] ["foo" "foo"]]]
                     (-> (mc/explain [:sequential [:tuple number? string?]] values)
                         (me/humanize)))
[nil nil [["should be a number"]]]
```

dpsutton commented on test/metabase/models/model_index_test.clj:
```diff
@@ -181,6 +215,7 @@
                                                                :value_ref value-ref})]
               (is (nil? error))
               ;; oracle returns BigDecimal ids so need `number?` rather than `int?`
```
fixed. also cleaned up an issue with the error message as well
